### ISSUE TITLE
Possible NPE when receiving a response (without a request)

### ISSUE THREAD

**Issue by jrhee17** [line/armeria#3055]
While upgrading to the most recent intellij (which automatically creates a breakpoint on NPE) I found this issue.

Our 0.99.8 armeria client connects to an endpoint via a load balancer which sometimes sends a 408 http response (without a http request)
https://tools.ietf.org/html/rfc7231#section-6.5.7

This results in an NPE at the following location:

https://github.com/line/armeria/blob/ba2601eb83e2072e1730ea9765bd332947b7b7c4/core/src/main/java/com/linecorp/armeria/client/Http1ResponseDecoder.java#L172

While I don't think our servers have had any issues due to this, I think it might make more sense to explicitly close the connection from the client side as well instead of silently ignoring the error.

**Comment by minwoox** [comment]
> a load balancer which sometimes sends a 408 http response (without a http request)

That's interesting. So the load balancer sends the 408 response even though it doesn't get any requests? so it means that the response is sent in the idle connections. right?

**Comment by jrhee17** [comment]
> so it means that the response is sent in the idle connections. right?

I haven't been able to exactly verify which connections were closed.

Given that the client is configured to use `h1c` with no `pipelining` and we don't use long polling in our business logic,
I was guessing it's probably due to health checks (we are using `HealthCheckService` with default values).

I can also share our code from a private channel if you would like to take a closer look

**Comment by jrhee17** [comment]
Related to #2900

**Comment by minwoox** [comment]
Ah, I forgot to reply to this. 😅 

> I was guessing it's probably due to health checks (we are using HealthCheck[Service][EndpointGroup] with default values).

Or Ping request? 😄 
I just wanted to make sure whether it's the problem of the load balancer or our client which might not correctly track the requests and responses

**Comment by jrhee17** [comment]
It's a little complicated, so uploaded a PR to illustrate what's happening.

1. The cause was definitely client request timeout, ping wasn't enabled.
2. Aborting a request immediately afterwards, and sending a 408 client request timeout causes this issue (without additional requests on the connection).
3. The first `HttpResponse` obj was swallowed after a NPE, the corresponding next `HttpContent` threw a `ProtocolViolationException`.
4. In our case, this log was visible when using `OrElseEndpointGroup` where the first was a `DynamicEndpointGroup`, and the second was a `StaticEndpointGroup`

**Comment by trustin** [comment]
@jrhee17 Would you mind elaborating on this:

> Aborting a request immediately afterwards, and sending a 408 client request timeout causes this issue (without additional requests on the connection).

Is this problem triggered by a ill-behaving load balancer that sends a response event before receiving a request or ..?

**Comment by jrhee17** [comment]
> a ill-behaving load balancer 

I guess some load balancers close a open connection without a request for a certain period of time.
I don't know if I would call it ill-behaving since I was able to reproduce this with a local haproxy with configurations similar to the link below (it's been a while, so I have to dig for the actual cfg file)
https://www.haproxy.com/blog/haproxy-and-http-errors-408-in-chrome/

**Comment by trustin** [comment]
I see. I guess what we can do possibly are one of the following:

- Close the connection when it was returned to the pool without making any requests.
- Just warn the protocol violation and close the connection.
  - We could log at debug level if status code is 408.

**Comment by jrhee17** [comment]
It's been a while since I took a look at this issue, so let me organize what I am thinking again.

---

I can still reproduce the warning logs - this can be done by copying the below code and running it
https://github.com/jrhee17/armeria/blob/a64d1e688f5c2378375e1f50fff29c4e43226f14/core/src/test/java/com/linecorp/armeria/client/HttpClientRequestTimeoutTest.java

**Explanation**

1. Create an armeria `h1c` client
2. Try to make a request and abort the request immediately - this is to ensure that a connection is established, but a request isn't sent.
3. Send a 408 request timeout response from the server side.
    - https://tools.ietf.org/html/rfc7231#section-6.5.7
4. The armeria client prints the following log:

```
22:19:50.566 [main] INFO  com.linecorp.armeria.common.Flags - com.linecorp.armeria.dumpOpenSslInfo: false (default)
22:19:50.729 [main] DEBUG c.l.armeria.common.util.SystemInfo - Java version: 14
22:19:51.913 [armeria-common-worker-nio-5-1] WARN  c.l.a.client.Http1ResponseDecoder - Unexpected exception:
com.linecorp.armeria.common.ProtocolViolationException: unexpected message type: io.netty.handler.codec.http.DefaultHttpContent (expected: io.netty.handler.codec.http.HttpResponse, channel: [id: 0x93474c4b, L:/127.0.0.1:64190 ! R:127.0.0.1/127.0.0.1:64189], resId: 1)
	at com.linecorp.armeria.client.Http1ResponseDecoder.failWithUnexpectedMessageType(Http1ResponseDecoder.java:266)
	at com.linecorp.armeria.client.Http1ResponseDecoder.channelRead(Http1ResponseDecoder.java:190)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.flush.FlushConsolidationHandler.channelRead(FlushConsolidationHandler.java:152)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:832)
22:20:02.919 [Thread-0] DEBUG c.l.armeria.client.ClientFactory - Closing the default client factories
22:20:02.923 [Thread-0] DEBUG c.l.armeria.client.ClientFactory - Closed the default client factories

Process finished with exit code 0
```

**Comment by ikhoon** [comment]
Fixed by #3060
